I/6113: Base table selection plugin

jodator opened the pull request:
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Introduce `TableSelection` plugin. Closes ckeditor/ckeditor5#6113.

---

### Additional information

- :tada: a `TableSelection` has matured since its first commit in August of 2018.
- most important things here are: `TableSelection` plugin itself and the introduced `MouseSelectionHandler` - the architecture of this feature is meant to have keyboard handler in a similar fashion
- the conversion uses the highlight mechanism known from the widget (fake selection, etc)
- if something is not working or not implemented - check the epick first if it is not reported - this PR only enables table selection with a mouse - there are no commands that would support this ATM

Reinmar commented:
Some bugs:

* Mouse down, mouse move outside the editor, move back inside, selection stopped working.
* Leaving the table with the mouse makes the selection non-fake so you suddenly see text selection. The selection stops working in this case too.

IMO – selection started within table should end inside the table. This means – that when you  temporarily leave the table, those mousemove events should be ignored (like nothing's happening)

Reinmar commented:
Missing feature:

* Pressing left/right/top/bottom should trim the selection to the start/end of the first/last range.
* Shift+arrows should allow extending/shrinking the selection (separate ticket).
* Typing doesn't work, but I see two issues here:
   * Probably typing will have to do the thing described in the first port (collapsing the selection).
   * Looking at the error, there's something wrong with selection conversion at this stage.

jodator commented:
@Reinmar  so for the bugs:

1. Fixed on PR
2. Moved to [UI/UX issue](https://github.com/ckeditor/ckeditor5/issues/6114#issuecomment-587423974) - it is a Chrome issue IMO as on FF it works OK

As for the missing features:

1. somehow fixed on PR
2. added comment int keyborad selection issue
3. I've created a bug issue for this at this needs some time to debug implement. There's a starting point though.

oleq commented on src/tableselection/mouseselectionhandler.js:
```diff
@@ -0,0 +1,162 @@
+import MouseEventsObserver from './mouseeventsobserver';
+
+/**
+ * A mouse selection handler for table selection.
```
```suggestion
 * A mouse selection handler class for the table selection.
```

oleq commented on src/tableselection/mouseselectionhandler.js:
```diff
@@ -0,0 +1,162 @@
+ */
+export default class MouseSelectionHandler {
+	/**
+	 * Creates instance of `MouseSelectionHandler`.
```
```suggestion
	 * Creates an instance of the `MouseSelectionHandler`.
```

oleq commented on src/tableselection/mouseselectionhandler.js:
```diff
@@ -0,0 +1,162 @@
+	 */
+	constructor( tableSelection, editing ) {
+		/**
+		 * Table selection.
```
```suggestion
		 * The table selection plugin instance.
```

oleq commented on src/tableselection/mouseselectionhandler.js:
```diff
@@ -0,0 +1,162 @@
+
+		/**
+		 * A flag indicating that the mouse selection is "active". A selection is "active" if it was started and not yet finished.
+		 * A selection can be "active" for instance if user moves a mouse over a table while holding a mouse button down.
```
```suggestion
		 * A selection can be "active", for instance, if a user moves a mouse over a table while holding a mouse button down.
```

oleq commented on src/tableselection/mouseselectionhandler.js:
```diff
@@ -0,0 +1,162 @@
+	}
+
+	/**
+	 * Handles starting a selection when "mousedown" event has table cell target.
```
```suggestion
	 * Handles starting a selection when "mousedown" event has table cell as a DOM target.
```

oleq commented on src/tableselection/mouseselectionhandler.js:
```diff
@@ -0,0 +1,162 @@
+	/**
+	 * Handles starting a selection when "mousedown" event has table cell target.
+	 *
+	 * If no table cell in event target it will clear previous selection.
```
```suggestion
	 * If there is no table cell in the event target, it will clear the previous selection.
```

oleq commented on src/tableselection/mouseselectionhandler.js:
```diff
@@ -0,0 +1,162 @@
+	}
+
+	/**
+	 * Handles updating table selection when "mousemove" event has a table cell target.
```
```suggestion
	 * Handles updating the table selection when the "mousemove" event has a table cell as a DOM target.
```

oleq commented on src/tableselection/mouseselectionhandler.js:
```diff
@@ -0,0 +1,162 @@
+	/**
+	 * Handles updating table selection when "mousemove" event has a table cell target.
+	 *
+	 * Does nothing if no table cell in event target or selection is not started.
```
```suggestion
	 * Does nothing if there is no table cell in event target or the selection has not been started yet.
```

oleq commented on src/tableselection/mouseselectionhandler.js:
```diff
@@ -0,0 +1,162 @@
+	}
+
+	/**
+	 * Handles ending (not clearing) table selection on "mouseup" event.
```
```suggestion
	 * Handles ending (not clearing) the table selection on the "mouseup" event.
```

oleq commented on src/tableselection/mouseselectionhandler.js:
```diff
@@ -0,0 +1,162 @@
+	/**
+	 * Handles ending (not clearing) table selection on "mouseup" event.
+	 *
+	 * Does nothing if selection is not started.
```
```suggestion
	 * Does nothing if the selection has not been started yet.
```

oleq commented on src/tableselection/mouseselectionhandler.js:
```diff
@@ -0,0 +1,162 @@
+	}
+
+	/**
+	 * Finds model table cell for given DOM event.
```
```suggestion
	 * Finds a model table cell for a given DOM event.
```

oleq replied:
Wasn't there a similar helper in the table cell property utils? (I'm not sure)

jodator replied:
I couldn't find anything that uses `toModelElement()` from a view element.

oleq commented on tests/_utils/utils.js:
```diff
@@ -322,11 +322,59 @@ export function assertTableStyle( editor, tableStyle, figureStyle ) {
 }
 
+/**
+ * Helper method for asserting selected table cells.
```
```suggestion
 * A helper method for asserting selected table cells.
```

oleq commented on tests/manual/tableselection.html:
```diff
@@ -0,0 +1,57 @@
+</style>
+
+<div id="editor">
+	<p>A table to test selection:</p>
```
This test's content is too homogenous to properly test the feature (as follow-ups has shown). We need more cases, for instance:
* with the content before and after the table,
* with a solitary table in the content,
* with a table that has some cells already split/merged,
* with a table that has links, block quotes and other content in some cells.

jodator replied:
I have something already for some of the points in the copy table fragment PR. I can cherry pick it here.

jodator replied:
But I'm not sure how to handle all these issues in one manual test. I've added only a complex table (which covers two of those issues).

oleq replied:
Create 2 editors? 😛

jodator replied:
But we have 4 or more cases :P 

But what do you think about loading fixtures on a button? One editor (I'd like to have one editor as copy-paste manual tests enhancement would require an editiable arae side by side to the editor.

jodator replied:
Moved to: https://github.com/ckeditor/ckeditor5/issues/6306

jodator commented on src/tableselection.js:
```diff
@@ -0,0 +1,254 @@
+	}
+
+	/**
+	 * Updates current table selection end element. Table selection is defined by #start and #end element.
```
I fixed it by expanding the docs.

jodator commented on src/tableselection.js:
```diff
@@ -0,0 +1,254 @@
+	/**
+	 * Starts a selection process.
+	 *
+	 * This method enables the table selection process.
```
This section was rewritten.

jodator commented on theme/tableediting.css:
```diff
@@ -8,3 +8,10 @@
  * **in the theme** when compiling the editor.
  */
+
+.ck-content .table table {
```
Please excuse my 2018 friend :) I didn't validate this when working on it - just changed the styles. I think tha this also should land in `TableSelection` and not in editing part, am I right?

jodator replied:
I'm not going to move this to theme-lark :cry: I would like to avoid `table-selection` branch handling on differnet repos. I'll add this remark to `UI/UI` issue though.

jodator commented on src/tableselection.js:
```diff
@@ -0,0 +1,254 @@
+		const startLocation = this._tableUtils.getCellLocation( this._startElement );
+		const endLocation = this._tableUtils.getCellLocation( this._endElement );
+
+		const startRow = startLocation.row > endLocation.row ? endLocation.row : startLocation.row;
```
:+1: This was also me from 2018 :wink:  but me from a few days ago didn't check that :disappointed:

oleq commented:
Post-review follow-ups:
* https://github.com/ckeditor/ckeditor5/issues/6302 (lack of background for selected cells)
* https://github.com/ckeditor/ckeditor5/issues/6301 (enter and backspace issues)
* https://github.com/ckeditor/ckeditor5/issues/6300 (selection is spilling out)
* https://github.com/ckeditor/ckeditor5/issues/6299 (selection lost if goes beyond the editor)
